File: js/src/stream/hash.ts
/**
 * SHA-256 state hash helpers for GS1.
 * 
 * Uses the Web Crypto API for SHA-256.
 */

import { canonicalizeLoose } from '../loose';
import { GValue } from '../types';

/**
 * Compute SHA-256 of the given bytes.
 * Works in both Node.js and browsers.
 */
export async function sha256(data: Uint8Array): Promise<Uint8Array> {
  if (typeof crypto !== 'undefined' && crypto.subtle) {
    // Browser or Node 15+
    const hash = await crypto.subtle.digest('SHA-256', data);
    return new Uint8Array(hash);
  }
  
  // Node.js fallback
  const { createHash } = await import('crypto');
  const hash = createHash('sha256').update(data).digest();
  return new Uint8Array(hash);
}

/**
 * Compute SHA-256 synchronously (Node.js only).
 */
export function sha256Sync(data: Uint8Array): Uint8Array {
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  const { createHash } = require('crypto');
  const hash = createHash('sha256').update(data).digest();
  return new Uint8Array(hash);
}

/**
 * Compute state hash using CanonicalizeLoose.
 * This is: sha256(CanonicalizeLoose(value))
 */
export async function stateHashLoose(value: GValue): Promise<Uint8Array> {
  const canonical = canonicalizeLoose(value);
  const encoder = new TextEncoder();
  return sha256(encoder.encode(canonical));
}

/**
 * Compute state hash synchronously (Node.js only).
 */
export function stateHashLooseSync(value: GValue): Uint8Array {
  const canonical = canonicalizeLoose(value);
  const encoder = new TextEncoder();
  return sha256Sync(encoder.encode(canonical));
}

/**
 * Compute state hash from raw bytes.
 */
export async function stateHashBytes(data: Uint8Array): Promise<Uint8Array> {
  return sha256(data);
}

/**
 * Verify that the current state hash matches the expected base.
 */
export function verifyBase(current: Uint8Array, expected: Uint8Array): boolean {
  if (current.length !== expected.length) return false;
  for (let i = 0; i < current.length; i++) {
    if (current[i] !== expected[i]) return false;
  }
  return true;
}

/**
 * Convert a 32-byte hash to lowercase hex string.
 */
export function hashToHex(h: Uint8Array): string {
  const hex = '0123456789abcdef';
  let result = '';
  for (let i = 0; i < h.length; i++) {
    result += hex[h[i] >> 4];
    result += hex[h[i] & 0x0f];
  }
  return result;
}

/**
 * Parse a 64-character hex string to a 32-byte hash.
 */
export function hexToHash(s: string): Uint8Array | null {
  // Strip optional prefix
  if (s.startsWith('sha256:')) {
    s = s.slice(7);
  }
  
  if (s.length !== 64) {
    return null;
  }
  
  const hash = new Uint8Array(32);
  for (let i = 0; i < 32; i++) {
    const hi = hexDigit(s.charCodeAt(i * 2));
    const lo = hexDigit(s.charCodeAt(i * 2 + 1));
    if (hi < 0 || lo < 0) {
      return null;
    }
    hash[i] = (hi << 4) | lo;
  }
  return hash;
}

function hexDigit(c: number): number {
  if (c >= 48 && c <= 57) return c - 48;      // 0-9
  if (c >= 97 && c <= 102) return c - 97 + 10; // a-f
  if (c >= 65 && c <= 70) return c - 65 + 10;  // A-F
  return -1;
}

File: go/stream/hash.go
package stream

import (
	"crypto/sha256"

	"github.com/Neumenon/glyph/go/glyph"
)

// StateHashLoose is the GS1 state hash: sha256(glyph.CanonJSON(value)), the
// one digest (SPEC-CANON.md §5), as 32 raw bytes. HashToHex of it equals
// glyph.Fingerprint(value). Identical in Python and JS.
func StateHashLoose(value *glyph.GValue) ([32]byte, error) {
	c, err := glyph.CanonJSON(value)
	if err != nil {
		return [32]byte{}, err
	}
	return sha256.Sum256(c), nil
}

// StateHashEmit computes the state hash using Emit (default emit).
// This is: sha256(Emit(decoded value))
//
// Use this when you need schema-aware output but don't have a schema
// for canonical packed encoding.
func StateHashEmit(value *glyph.GValue) [32]byte {
	emitted := glyph.Emit(value)
	return sha256.Sum256([]byte(emitted))
}

// StateHashBytes computes SHA-256 of raw bytes.
// Use this when you already have canonical bytes.
func StateHashBytes(data []byte) [32]byte {
	return sha256.Sum256(data)
}

// VerifyBase checks if the current state hash matches the expected base.
func VerifyBase(current, expected [32]byte) bool {
	return current == expected
}

// HashToHex converts a 32-byte hash to lowercase hex string.
func HashToHex(h [32]byte) string {
	const hextable = "0123456789abcdef"
	var buf [64]byte
	for i, b := range h {
		buf[i*2] = hextable[b>>4]
		buf[i*2+1] = hextable[b&0x0f]
	}
	return string(buf[:])
}

// HexToHash parses a 64-character hex string to a 32-byte hash.
func HexToHash(s string) ([32]byte, bool) {
	var h [32]byte
	if len(s) != 64 {
		return h, false
	}
	for i := 0; i < 32; i++ {
		hi := hexDigit(s[i*2])
		lo := hexDigit(s[i*2+1])
		if hi < 0 || lo < 0 {
			return h, false
		}
		h[i] = byte(hi<<4 | lo)
	}
	return h, true
}

func hexDigit(c byte) int {
	switch {
	case c >= '0' && c <= '9':
		return int(c - '0')
	case c >= 'a' && c <= 'f':
		return int(c - 'a' + 10)
	case c >= 'A' && c <= 'F':
		return int(c - 'A' + 10)
	default:
		return -1
	}
}
